samples/boot/oauth2resourceserver-multitenancy/src/main/java/sample/OAuth2ResourceServerSecurityConfiguration.java

```diff
@@ -0,0 +1,83 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sample;
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.security.authentication.AuthenticationDetailsSource;
+import org.springframework.security.authentication.AuthenticationManager;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.oauth2.jwt.JwtDecoder;
+import org.springframework.security.oauth2.jwt.JwtProcessors;
+import org.springframework.security.oauth2.jwt.NimbusJwtDecoder;
+import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationProvider;
+import org.springframework.security.oauth2.server.resource.authentication.OAuth2IntrospectionAuthenticationProvider;
+
+/**
+ * @author Josh Cummings
+ */
+@EnableWebSecurity
+public class OAuth2ResourceServerSecurityConfiguration extends WebSecurityConfigurerAdapter {
+
+	@Value("${spring.security.oauth2.resourceserver.jwt.jwk-set-uri}")
+	String jwkSetUri;
+
+	@Value("${spring.security.oauth2.resourceserver.opaque.introspection-uri}")
+	String introspectionUri;
+
+	@Override
+	protected void configure(HttpSecurity http) throws Exception {
+		// @formatter:off
+		http
+			.authorizeRequests()
+				.antMatchers("/**/message/**").hasAuthority("SCOPE_message:read")
+				.anyRequest().authenticated()
+				.and()
+			.oauth2ResourceServer()
+				.authenticationDetailsSource(authenticationDetailsSource());
+		// @formatter:on
+	}
+
+	@Bean
+	AuthenticationManager multitenantAuthenticationManager() {
+		return authentication -> {
+			String tenantId = (String) authentication.getDetails();
+			if ("tenantOne".equals(tenantId)) {
+				return jwt().authenticate(authentication);
+			} else if ("tenantTwo".equals(tenantId)){
+				return opaque().authenticate(authentication);
+			}
+			throw new IllegalArgumentException("Unknown tenant");
+		};
+	}
+
+	JwtAuthenticationProvider jwt() {
+		JwtDecoder jwtDecoder = new NimbusJwtDecoder(JwtProcessors.withJwkSetUri(this.jwkSetUri).build());
+		return new JwtAuthenticationProvider(jwtDecoder);
+	}
+
+	OAuth2IntrospectionAuthenticationProvider opaque() {
+		return new OAuth2IntrospectionAuthenticationProvider(this.introspectionUri, "client", "secret");
+	}
+
+	AuthenticationDetailsSource<HttpServletRequest, ?> authenticationDetailsSource() {
+		return request -> request.getPathInfo().split("/")[1];
```

Comment:
@jzheaux I don't think it makes sense to use `AuthenticationDetailsSource` for extracting the tenant info. In this example, `AuthenticationManager` indirectly relies on `HttpServletRequest` which is not the intended design for `AuthenticationManager` as it's not meant to be tied to the web-tier - it should work on a desktop application for example.

Also, the tenant info may come from a request parameter so I'm not sure how this would work with the current setup.

Either way, I feel we need to extract the tenant info in the web-tier and populate it with a new impl of  `Authentication` that is passed into `AuthenticationManager`. We may even consider a new type that is populated in a `AuthenticationDetailsSource`? I just think we should leave the extracting of the tenant info in the web-tier - `Filter` or something similar.

Reply:
> In this example, `AuthenticationManager` indirectly relies on `HttpServletRequest`

What do you mean that it is indirectly reliant? By that same logic, isn't it also indirectly reliant on `HttpServletRequest` because of the bearer token resolver? No web-ness makes it down into the authentication manager. We just have a tenantId that so happens to be the path.

This also happens on the reactive side. The `AuthenticationWebFilter` calls a `ServerAuthenticationConverter`, which takes a `request` and returns an `Authentication`, which is then passed to a manager.

> We may even consider a new type that is populated in a `AuthenticationDetailsSource

A stronger type may be in order, as you said, though. Calling `setDetails(Object)` doesn't offer any type safety, nor any direction on what ought to go there. However, I don't think that I'd want to enforce it at the setter level. For example, I wouldn't want to do this:

```java
public void setAuthenticationDetailsSource
    (AuthenticationDetailsSource<HttpServletRequest, Tenant>)
```

Since there are certainly other scenarios where the user would like to provide details in the `Authentication` object.

> I just think we should leave the extracting of the tenant info in the web-tier - `Filter` or something similar.

Just to be clear, this is already technically the case since the filter calls `source.buildDetails` before invoking the authentication manager. But, I'm thinking you are calling out something more nuanced than that.

Reply:
@jzheaux Regarding my comment

> `AuthenticationManager` indirectly relies on `HttpServletRequest` which is not the intended design...

Please ignore as I misinterpreted the implementation on my initial look. All seems fine. Actually, I like the fact that you're leveraging `Authentication.details` to store the tenant info. This allows for an easy integration.

I'm wondering if it makes sense to leverage (or extend) `WebAuthenticationDetailsSource` and store the tenant info in `WebAuthenticationDetails` or subclass of it?